tests/testthat/test-runner.R

```diff
@@ -359,10 +359,87 @@ test_that("status missing ID", {
     status = "missing",
     version = NULL,
     output = NULL,
-    task_position = 0
+    queue = list()
   ))
 })
 
+test_that("check_timeout kills timed out reports", {
+  testthat::skip_on_cran()
+  skip_on_appveyor()
+  skip_on_windows()
+  skip_if_no_redis()
+  path <- orderly_prepare_orderly_example("demo")
+  runner <- orderly_runner(path, workers = 2)
+
+  key1 <- runner$submit_task_report("minimal", timeout = 0)
+  key2 <- runner$submit_task_report("slow10", timeout = 0)
+  ## Sleep for a bit to allow "minimal" report to post TASK_COMPLETE message
+  Sys.sleep(3)
+  msg <- capture_messages(killed <- runner$check_timeout())
+  task2 <- get_task_id_key(runner, key2)
+  expect_equal(killed, task2)
+  expect_equal(msg, sprintf("Successfully killed '%s', exceeded timeout of 0\n",
+                            task2))
+})
+
+test_that("check_timeout doesn't kill reports with long timeout", {
+  testthat::skip_on_cran()
+  skip_on_appveyor()
+  skip_on_windows()
+  skip_if_no_redis()
+  path <- orderly_prepare_orderly_example("demo")
+  runner <- orderly_runner(path, workers = 2)
+
+  key1 <- runner$submit_task_report("minimal", timeout = 20)
+  key2 <- runner$submit_task_report("slow10", timeout = 20)
+  Sys.sleep(1) ## Sleep so both reports get started
+  msg <- capture_messages(killed <- runner$check_timeout())
+  expect_equal(killed, character(0))
+  expect_equal(msg, character(0))
+})
+
+test_that("check_timeout returns NULL if no reports being run", {
+  testthat::skip_on_cran()
+  skip_on_appveyor()
+  skip_on_windows()
+  skip_if_no_redis()
+  path <- orderly_prepare_orderly_example("interactive", testing = TRUE)
+  runner <- orderly_runner(path)
+  msg <- capture_messages(killed <- runner$check_timeout())
+  expect_null(killed)
+  expect_equal(msg, character(0))
+})
+
+test_that("check_timeout prints message if fails to kill a report", {
+  testthat::skip_on_cran()
+  skip_on_appveyor()
+  skip_on_windows()
+  skip_if_no_redis()
+  path <- orderly_prepare_orderly_example("interactive", testing = TRUE)
+  runner <- orderly_runner(path)
+
+  ## Setup mock queue with realistic worker log
+  key1 <- runner$submit_task_report("interactive", timeout = 0)
+  Sys.sleep(0.5) ## Sleep to wait for runner to pick up task
+  logs <- runner$queue$worker_log_tail()
+
+  queue <- runner$queue
+  mock_queue <- list(
```

Comment:
This is pretty horrible too. I want to create a test to cover the case where in between getting the logs and the queue saying that task A can be killed it then completing meaning that the call to cancel the task run fails. I could alternatively create a helper function for returning the logs and then mock that out but that feels like a unnecessary extra function, but perhaps is justified if it makes test writing easier. Any thoughts?

Reply:
this is a hard thing to do without exposing a lot of details

Can you add this comment as a code comment so we know why the test exists and if things work emperically and the design as settled down we can look at making these tests more robust as a separate job?